src/lib/pubkey/kyber/kyber_common/kyber.cpp

```diff
@@ -517,17 +515,49 @@ class Polynomial
        * Run rejection sampling on uniform random bytes to generate uniform
        * random integers mod q.
        */
-      static Polynomial sample_rej_uniform(std::unique_ptr<StreamCipher> xof)
+      static Polynomial sample_rej_uniform(Kyber_XOF& xof)
          {
+         class XOF_Reader
+            {
+            public:
+               XOF_Reader(Kyber_XOF& xof) :
+                  m_xof(xof),
+                  m_buf(),
+                  m_buf_pos(m_buf.size())
+                  {
+                  }
+
+               std::array<uint8_t, 3> next_3_bytes()
+                  {
+                  if(m_buf_pos == m_buf.size())
+                     {
+                     m_xof.write_output(m_buf.data(), m_buf.size());
+                     m_buf_pos = 0;
+                     }
+
```

Comment:
Maybe to be on the safe side:

```suggestion
BOTAN_ASSERT_NOMSG(m_buf_pos + 3 <= m_buf.size());
```

Reply:
:+1: